osu.Server.Queues.ScoreStatisticsProcessor/Processors/PlayTimeProcessor.cs

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Dapper;
+using JetBrains.Annotations;
+using MySqlConnector;
+using osu.Game.Rulesets;
+using osu.Game.Rulesets.Mods;
+using osu.Server.Queues.ScoreStatisticsProcessor.Models;
+
+namespace osu.Server.Queues.ScoreStatisticsProcessor.Processors
+{
+    /// <summary>
+    /// Increment total user play time.
+    /// </summary>
+    [UsedImplicitly]
+    public class PlayTimeProcessor : IProcessor
+    {
+        public void RevertFromUserStats(SoloScoreInfo score, UserStats userStats, int previousVersion, MySqlConnection conn, MySqlTransaction transaction)
+        {
+            if (previousVersion >= 6)
+                userStats.total_seconds_played -= getPlayLength(score, conn, transaction);
+        }
+
+        public void ApplyToUserStats(SoloScoreInfo score, UserStats userStats, MySqlConnection conn, MySqlTransaction transaction)
+        {
+            if (score.ended_at == null)
+                throw new InvalidOperationException("Attempting to increment play time when score was never finished.");
+
+            userStats.total_seconds_played += getPlayLength(score, conn, transaction);
+        }
+
+        public void ApplyGlobal(SoloScoreInfo score, MySqlConnection conn)
+        {
+        }
+
+        private static int getPlayLength(SoloScoreInfo score, MySqlConnection conn, MySqlTransaction transaction)
+        {
+            Debug.Assert(score.started_at != null);
+            Debug.Assert(score.ended_at != null);
+
+            // to ensure sanity, first get the maximum time feasible from the beatmap's length
+            double totalLengthSeconds = conn.QueryFirstOrDefault<double>("SELECT total_length FROM osu_beatmaps WHERE beatmap_id = @beatmap_id", score, transaction);
+
+            Ruleset ruleset = ScoreStatisticsProcessor.AVAILABLE_RULESETS.Single(r => r.RulesetInfo.ID == score.ruleset_id);
+
+            var rateAdjustMods = score.mods.Select(m => m.ToMod(ruleset)).OfType<ModRateAdjust>().ToArray();
```

Comment:
I was going to suggest using `IApplicableToRate` because the whole `ModTimeRamp` subtree is excluded here, but that can be done later because it's a bit of a weird one.  
It would mean that you have to decide on a specific time to take the rate at and I'm not able to mentally prove if taking `base_duration / 2` as that time gives you the correct rate to approximate length in this case.

This is a fallback case if I understand it anyway so at worst it will be overestimating length by a bit.

Reply:
Yeah, in theory I would hope we can trust start/end time due to other means of verification.